npm/podController.go

```diff
@@ -0,0 +1,623 @@
+			utilruntime.HandleError(fmt.Errorf("expected string in workqueue but got %#v", obj))
+			return nil
+		}
+		// Run the syncHandler, passing it the namespace/name string of the
+		// Pod resource to be synced.
+		// (TODO): may consider using "c.queue.AddAfter(key, *requeueAfter)" according to error type later
+		if err := c.syncPod(key); err != nil {
+			// Put the item back on the workqueue to handle any transient errors.
+			c.workqueue.AddRateLimited(key)
+			return fmt.Errorf("error syncing '%s': %s, requeuing", key, err.Error())
+		}
+		// Finally, if no error occurs we Forget this item so it does not
+		// get queued again until another change happens.
+		c.workqueue.Forget(obj)
+		log.Logf("Successfully synced '%s'", key)
+		return nil
+	}(obj)
+
+	if err != nil {
+		utilruntime.HandleError(err)
+		return true
+	}
+
+	return true
+}
+
+// syncPod compares the actual state with the desired, and attempts to converge the two.
+func (c *podController) syncPod(key string) error {
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return nil
+	}
+
+	// Get the Pod resource with this namespace/name
+	pod, err := c.podLister.Pods(namespace).Get(name)
+	// (TODO): Reduce scope of lock later
+	c.npMgr.Lock()
+	defer c.npMgr.Unlock()
+
+	if err != nil {
+		if errors.IsNotFound(err) {
+			log.Logf("pod %s not found, may be it is deleted", key)
+			cachedNpmPodObj, exist := c.npMgr.PodMap[key]
+			// if the npmPod does not exists, we do not need to clean up process and retry it
+			if !exist {
+				return nil
+			}
+
+			// Found the npmPod object from PodMap local cache and start cleaning up processes
+			err = c.cleanUpDeletedPod(cachedNpmPodObj.getPodObjFromNpmPodObj())
+			if err != nil {
+				// need to retry this cleaning-up process
+				metrics.SendErrorLogAndMetric(util.PodID, "Error: %v when pod is not found", err)
+				return fmt.Errorf("Error: %v when pod is not found\n", err)
+			}
+			return err
+		}
+
+		return err
+	}
+
+	// If newPodObj status is either corev1.PodSucceeded or corev1.PodFailed or DeletionTimestamp is set, start clean-up the lastly applied states.
+	if isCompletePod(pod) {
+		if err = c.cleanUpDeletedPod(pod); err != nil {
+			metrics.SendErrorLogAndMetric(util.PodID, "Error: %v when pod is in completed state.", err)
+			return fmt.Errorf("Error: %v when when pod is in completed state.\n", err)
+		}
+		return nil
+	}
+
+	err = c.syncAddAndUpdatePod(pod)
+	if err != nil {
+		metrics.SendErrorLogAndMetric(util.PodID, "Error: Failed to sync pod due to %v", err)
+		return fmt.Errorf("Failed to sync pod due to %v\n", err)
+	}
+
+	return nil
+}
+
+func (c *podController) syncAddedPod(podObj *corev1.Pod) error {
+	npmPodObj := newNpmPod(podObj)
+	podNs := util.GetNSNameWithPrefix(podObj.Namespace)
+	podKey, _ := cache.MetaNamespaceKeyFunc(podObj)
```

Comment:
utilruntime.HandleError ?

Reply:
This seems a convention in K8s controller implementation. So we also follow it. In high level, this method is preferable to logging the error (https://pkg.go.dev/k8s.io/apimachinery/pkg/util/runtime#HandleError)